### PR TITLE
Attempt to fix the login url comparison

### PR DESCRIPTION
**Pull request by burzum**
Refs https://github.com/cakephp/authentication/issues/137

### CODE REVIEW

**Comment by markstory** [comment]
Given that we can only confidently have string URLs to compare with loginAction, should we accept a list of URLs instead of a routing array? Something like 

```php
['loginAction' => ['/users/login', '/en/login']]
```

We could also support syntax like `/*`  for URL patterns with trailing arguments. This is a bit of a departure from AuthComponent, but helps decouple this from Routing.

**Comment by burzum** [comment]
@markstory could we support at least regex as well then?

We have 18 languages and your suggestion will enforce that somebody doesn't forgot to add a new language there when needed.

I could change the code in a way it supports the router but only when it's present, so this package won't have a hard dependency on it.

**Comment by dereuromark** [comment]
Can we have a few test cases for this scenario maybe?

**Comment by burzum** [comment]
@dereuromark before I waste time and it gets rejected *after* I wrote a test I would prefer to get an opinion if this is going to be OK?

https://gist.github.com/burzum/efe49d06fc928dec4955ad2291c7be81

File: src/Authenticator/FormAuthenticator.php [hunk 1]
```diff
@@ -100,12 +105,26 @@ protected function _checkLoginUrl(ServerRequestInterface $request)
...
 
         if (!empty($loginUrl)) {
-            if (is_array($loginUrl)) {
-                $loginUrl = Router::url(['_base' => false] + $loginUrl);
+            $requestUrl = Router::parseRequest($request);
+            unset($request['_matchedRoute']);
+
+            if (is_string($loginUrl)) {
+                $loginUrl = Router::parseRequest((new ServerRequest([
+                    'uri' => $loginUrl
```
Comment (markstory) [inline src/Authenticator/FormAuthenticator.php]: I think you want `url` not `uri`.

File: src/Authenticator/FormAuthenticator.php [hunk 2]
```diff
@@ -100,12 +105,26 @@ protected function _checkLoginUrl(ServerRequestInterface $request)
         $loginUrl = $this->getConfig('loginUrl');
 
         if (!empty($loginUrl)) {
-            if (is_array($loginUrl)) {
-                $loginUrl = Router::url(['_base' => false] + $loginUrl);
+            $requestUrl = Router::parseRequest($request);
+            unset($request['_matchedRoute']);
+
+            if (is_string($loginUrl)) {
+                $loginUrl = Router::parseRequest((new ServerRequest([
```
Comment (markstory) [inline src/Authenticator/FormAuthenticator.php]: Extra `(` here.

**Comment by robertpustulka** [comment]
Regex seems ok as well the explicit array support. Relying on router would probably mean that the authentication middleware must be always added after the routing middleware so the routes are build before the check made by the authenticator.

File: src/Authenticator/FormAuthenticator.php [hunk 3]
```diff
@@ -64,8 +64,9 @@ public function authenticate(ServerRequestInterface $request, ResponseInterface
         if (!$this->_checkLoginUrl($request)) {
             $loginUrl = $this->getConfig('loginUrl');
             if (is_array($loginUrl)) {
-                $loginUrl = Router::url($loginUrl, true);
+                $loginUrl = \Cake\Routing\Router::url($loginUrl, true);
             }
+            dd($loginUrl);
```
Comment (dereuromark) [inline src/Authenticator/FormAuthenticator.php]: dd

File: src/Authenticator/FormAuthenticator.php [hunk 4]
```diff
@@ -12,7 +12,8 @@
  */
 namespace Authentication\Authenticator;
 
-use Cake\Routing\Router;
+use Cake\Http\ServerRequest;
```
Comment (robertpustulka) [inline src/Authenticator/FormAuthenticator.php]: Looks like `cakephp/cakephp` dependency is required for this class to be loaded.

I'm still not convinced that this should rely on HTTP and/or Routing packages. Could we just support string urls as @markstory suggested? `loginUrl` option could accept an explicit url string (like `/login`), an array of urls (`['/login', '/users/login']`) or a regex (`%^/[a-z]+/login$%`).

Comment (burzum) [reply src/Authenticator/FormAuthenticator.php]: @robertpustulka would you mind doing the change then? I've explicitly asked before I started on the change but didn't get any clear yes or no on this... So please change it as you like until it fits what you want.

**Comment by burzum** [comment]
@markstory @robertpustulka any idea why the URL is wrong in the test? The routing setup seems to be right?

File: src/Authenticator/FormAuthenticator.php (hunk 4 above)
Comment (robertpustulka) [reply src/Authenticator/FormAuthenticator.php]: @burzum Sorry, I may have not been clear on this here: https://github.com/cakephp/authentication/pull/146#issuecomment-346813351

I’ll take a look on this tomorrow.

File: src/Authenticator/FormAuthenticator.php [hunk 5]
```diff
@@ -100,14 +105,76 @@ protected function _checkLoginUrl(ServerRequestInterface $request)
...
         if (!empty($loginUrl)) {
-            if (is_array($loginUrl)) {
-                $loginUrl = Router::url(['_base' => false] + $loginUrl);
-                $this->setConfig('loginUrl', $loginUrl);
+            if (is_string($loginUrl) && $this->_compareStringUrl($loginUrl, $request->getUri()->getPath())) {
+                return true;
             }
 
-            return strcasecmp($request->getUri()->getPath(), $loginUrl) === 0;
+            return $this->_compareArrayUrl($loginUrl, $request);
```
Comment (markstory) [inline src/Authenticator/FormAuthenticator.php]: Do we need array URL support? I'd like to believe that simple string URLs and regex patterns should cover the bulk of use cases.

File: src/Authenticator/FormAuthenticator.php [hunk 6]
```diff
@@ -100,14 +105,76 @@ protected function _checkLoginUrl(ServerRequestInterface $request)
...
+    {
+        if (!is_string($loginUrl)) {
+            throw new InvalidArgumentException();
+        }
+
+        if ($loginUrl === $requestUrl) {
+            return true;
+        }
+
+        $regex = $this->getConfig('loginUrlRegex');
```
Comment (markstory) [inline src/Authenticator/FormAuthenticator.php]: Having a separate config option is a good idea :+1: Better than using `preg:` prefix as we've done in other places in the past.

File: src/Authenticator/FormAuthenticator.php (hunk 5 above)
Comment (robertpustulka) [reply src/Authenticator/FormAuthenticator.php]: > Do we need array URL support?

I don't think we do. 

I can handle string URL approach if you @burzum and @markstory don't mind.

**Pull request closed without merging**